### PR TITLE
Native table update

### PR DESCRIPTION
**Pull request by jaredsnyder**
Changes:
- add an additional factor to forecast position-level inventory
- fix some names, specifically replacing "newtab_impressions" with "newtab_visits"

Validation:
First, validate that the position-level impressions accurately reflect actuals.  This was done by running the forecast to predict 2024 data based on past data.  The results are shown for the largest markets (US, UK and DE) below, with "cumulative" showing the metrics when calculated across the sum of impressions for the positions that can contain sponsored stories.  These positions are shown in red:

image
image
image

The forecast clearly need for optimization, these results are comparable to the high-single digit MAPE calculated for the newtab impression forecast (see: https://github.com/mozilla/moz-forecasting/pull/10)

Next, validate that the forecasted values make sense relative to the SPOCs sold.  This is done by looking at the ratio of inventory to SPOC impressions.  It should never be greater than one, but generally be >.9 for the most active countries, especially the US:

image
image
image

The UK (GB) is fairly low, but DE and US look okay.  Will follow up with Laura to make sure the low GB numbers make sense

### CODE REVIEW

File: moz_forecasting/native/README.md [hunk 1]
```diff
@@ -1,16 +1,20 @@
 # Native Forecast
 
 ## Overview
-This flow forecasts the number of newtab impressions for which sponsored stories are enabled.  The output of this forecast is used to forecast SPOC inventory by multiplying the result by the number of SPOCs per page (currently 1, because most users only see the top ad, which creates a confusing scenario where the spoc inventory forecast has the samen numerical value as the newtab impression forecast.) 
+This flow forecasts SPOC inventory, meaning the number of **total** (as oppposed to just sponsored) pocket impressions by tile.  This value is the maximum available number of impressions that could be sold as SPOCs.
```
Comment (sergiosonline) [inline moz_forecasting/native/README.md:4]: Clarifying that SPOC impressions != all Native impressions. SPOCs refer to ads in Native--https://mozilla-hub.atlassian.net/wiki/spaces/MA1/pages/328565119/Glossary. Can you clarify whether you're forecasting all impressions across all (sponsored and non-sponsored) positions, or only sponsored (SPOC) impressions.

File: moz_forecasting/native/README.md [hunk 2]
```diff
@@ -1,16 +1,20 @@
...
 
-The output table is indexed by country, month (`submission_month`), with the forecast in the `native_forecast` column
+The output table is indexed by country, month (`submission_month`), and position with the forecast in the `forecast_spoc_inventory` column
 
 Currently the data only runs on Firefox desktop
 
 ### `inventory_forecast` derivation
-The idea behind the inventory forecast is to execute the following steps:
+![inventory forecast flow diagram](native.drawio.png)
+The idea behind the inventory forecast is to execute the following steps. illustarted in the diagram above:
```
Comment (sergiosonline) [inline moz_forecasting/native/README.md:12]:
```suggestion
The idea behind the inventory forecast is to execute the following steps, illustrated in the diagram above:
```

File: moz_forecasting/native/README.md [hunk 3]
```diff
@@ -1,16 +1,20 @@
...
+- **turn that into country-level newtab visits with sponsored stories by multiplying a factor representing the ratio of visits to the dau by country**: The country-level inventory by month is obtained from the `mozdata.telemetry.newtab_clients_daily` table.  Visits are counted by month and country when the `pocket_sponsored_stories_enabled` and `pocket_enabled` flags are true.  The ratio is calculated by month and rolled up to the country level by averaging over one year.  It is first stored in `self.impressions_to_newtab_with_spocs_factor["ratio_newtab_visits_with_spocpocket_to_dou"]`.
+- **turn this into SPOC inventory by multiplying by a factor representing the number of impressions per newtab visit by position and country**: The by-country, by-position value is obtained from `mozdata.telemetry.newtab_visits`.  Visits are counted by position, month and country when the `pocket_sponsored_stories_enabled` and `pocket_enabled` flags are true. It is averaged over 12 months.  It is stored in `self.spocs_to_newtab_visits_factor["ratio_pocket_impressions_to_newtab_visits"]`
```
Comment (sergiosonline) [inline moz_forecasting/native/README.md:15]: Correct this based on the above comment around Native impressions != SPOC impressions.

File: moz_forecasting/native/flow.py [hunk 4]
```diff
@@ -283,22 +325,22 @@ def get_dau_forecast_by_country(self):
                 * dau_forecast_by_country["dau_forecast"]
             )
         self.dau_forecast_by_country = dau_forecast_by_country
-        self.next(self.get_newtab_impressions)
+        self.next(self.get_newtab_visits)
 
     @step
-    def get_newtab_impressions(self):
+    def get_newtab_visits(self):
```
Comment (sergiosonline) [inline moz_forecasting/native/flow.py:331]: Thanks for updating naming here. Much clearer.

File: moz_forecasting/native/flow.py [hunk 5]
```diff
@@ -310,34 +352,88 @@ def get_newtab_impressions(self):
...
             ]
-            .groupby("country", as_index=False)
+            .groupby("country_code", as_index=False)
+            .mean()
+        )
+        self.next(self.get_pocket_impressions)
+
+    @step
+    def get_pocket_impressions(self):
+        """Get ratio of pocket impressions by qualified newtab visits."""
```
Comment (sergiosonline) [inline moz_forecasting/native/flow.py:391]: Can you add a few words either here in the docstring or in the README.md around the exact criteria qualifying newtab visits for this forecast?

File: moz_forecasting/native/README.md (hunk 1 above)
Comment (sergiosonline) [reply moz_forecasting/native/README.md:4]: You may also need to update the flow diagram to account for this disambiguation.

Comment (jaredsnyder) [reply moz_forecasting/native/README.md:4]: SPOC **inventory** is all native impressions

File: moz_forecasting/native/flow.py (hunk 5 above)
Comment (jaredsnyder) [reply moz_forecasting/native/flow.py:391]: Added a comment in the previous step to clarify

File: moz_forecasting/native/README.md (hunk 1 above)
Comment (jaredsnyder) [reply moz_forecasting/native/README.md]: I rewrote this paragraph to make it more clear

File: moz_forecasting/native/README.md [hunk 6]
```diff
@@ -1,16 +1,23 @@
 # Native Forecast
 
 ## Overview
-This flow forecasts the number of newtab impressions for which sponsored stories are enabled.  The output of this forecast is used to forecast SPOC inventory by multiplying the result by the number of SPOCs per page (currently 1, because most users only see the top ad, which creates a confusing scenario where the spoc inventory forecast has the samen numerical value as the newtab impression forecast.) 
+This flow forecasts SPOC (Sponsored POCket) inventory, meaning the maximum possible impressions available to sell and display as sponsored stories.  The maximum number of possible SPOC impressions is the same as the number of **total** pocket impressions (as oppposed to just sponsored) by tile.  To put it another way, if 100% of the inventory was sold the number of SPOC impressions would equal the inventory, but if less than all of the inventory was sold the unsold inventory would be filled with organic stories.  Thus, the inventory will be the total number of pocket impressions.
```
Comment (sergiosonline) [inline moz_forecasting/native/README.md:4]:
```suggestion
This flow forecasts SPOC (Sponsored POCket) inventory, meaning the maximum possible impressions available to sell and display as sponsored stories **in the relevant ad positions (2, 6, 8, etc.)**.  The maximum number of possible SPOC impressions is the same as the number of **total** pocket impressions (as oppposed to just sponsored) by tile.  To put it another way, if 100% of the inventory was sold the number of SPOC impressions would equal the inventory, but if less than all of the inventory was sold the unsold inventory would be filled with organic stories.  Thus, the inventory will be the total number of pocket impressions.
```

File: moz_forecasting/native/README.md [hunk 7]
```diff
@@ -1,16 +1,23 @@
...
+- **turn this into SPOC inventory by multiplying by a factor representing the number of impressions per newtab visit by position and country**: The by-country, by-position value is obtained from `mozdata.telemetry.newtab_visits`.  Visits are counted by position, month and country when the `pocket_sponsored_stories_enabled` and `pocket_enabled` flags are true. It is averaged over 12 months.  It is stored in `self.spocs_to_newtab_visits_factor["ratio_pocket_impressions_to_newtab_visits"]`
+
+## Scheduled Production Run
+The monthly update it scheduled to run 1 AM UTC on the 3rd of every month. The deployment can be viewed in Outerbounds [here](https://ui.desertowl.obp.outerbounds.com/dashboard/flows/p/revenue/nativeforecast.prod.nativeforecastflow).  The deployment is updated automatically via the circleci job `update_native_scheduled_job` whenever a commit is made into main.
 
 ## History and Changes
+Code was updated to calculate SPOC inventory by position.
```
Comment (sergiosonline) [inline moz_forecasting/native/README.md:21]:
```suggestion
Code was updated to calculate Native SPOC inventory by position.
```

**Review by sergiosonline — COMMENTED** [review]
Added terminology, but looking better.

**Review by sergiosonline — APPROVED** [review]
leggoooo

**Pull request merged**